File: backend/routes/bets.js
// routes/bets.js
const express = require('express');
const prisma = require('../prismaClient');
const { authenticateUser } = require('../middleware/auth');

module.exports = (io, engine) => {
  const r = express.Router();

  // Apply auth middleware to all bet routes
  r.use(authenticateUser);

  r.post('/place', async (req, res) => {
    try {
      const { matchId, pick, odds, amount, currency } = req.body;
      const user = req.user; // from authenticateUser middleware

      if (!matchId || !pick || !odds || !amount) {
        return res.status(400).json({ error: 'Missing fields' });
      }

      // Check wallet balance
      const wallet = await prisma.wallet.findUnique({ where: { userId: user.id } });
      if (!wallet || wallet.balanceAvailable < amount) {
        return res.status(400).json({ error: 'Insufficient balance' });
      }

      // Deduct balance and create bet in transaction
      const bet = await prisma.$transaction(async (tx) => {
        // Deduct from wallet
        await tx.wallet.update({
          where: { userId: user.id },
          data: { balanceAvailable: { decrement: amount } }
        });

        // Record transaction
        await tx.walletTransaction.create({
          data: {
            walletId: wallet.walletId,
            transactionType: 'bet_placed',
            amount: amount,
            balanceBefore: wallet.balanceAvailable,
            balanceAfter: wallet.balanceAvailable - amount,
            transactionStatus: 'completed'
          }
        });

        // Place bet
        return await tx.bet.create({
          data: {
            userId: user.id,
            matchId: BigInt(matchId),
            betType: 'match_winner', // Assuming standard for now
            betAmount: amount,
            odds: odds,
            potentialPayout: Math.round(amount * odds),
            betStatus: 'pending'
          }
        });
      });

      // Serialization for BigInt
      const serializedBet = { ...bet, betId: bet.betId.toString(), userId: bet.userId.toString(), matchId: bet.matchId.toString() };

      res.json({ ok: true, bet: serializedBet });
    } catch (e) {
      console.error('[BETS PLACE]', e.message);
      res.status(500).json({ ok: false, error: 'Failed to place bet' });
    }
  });

  r.get('/user/:userId', async (req, res) => {
    try {
      const { userId } = req.params;
      
      // Basic check to ensure users only fetch their own bets unless admin (omitted for simplicity, but req.user.id can be checked)
      if (req.user.id.toString() !== userId) {
         return res.status(403).json({ error: 'Forbidden: Can only view your own bets' });
      }

      const userBets = await prisma.bet.findMany({
        where: { userId: BigInt(userId) },
        orderBy: { placedAt: 'desc' }
      });

      // Convert BigInt to strings
      const serializedBets = userBets.map(b => ({
        ...b,
        betId: b.betId.toString(),
        userId: b.userId.toString(),
        matchId: b.matchId.toString(),
      }));

      res.json({ bets: serializedBets });
    } catch (e) {
      console.error('[BETS GET]', e.message);
      res.status(500).json({ ok: false, error: 'Failed to get bets' });
    }
  });

  return r;
};

File: backend/routes/bets.go
package routes

import (
	"context"
	"fmt"
	"time"

	"github.com/gofiber/fiber/v2"

	"betelite-go/db"
	"betelite-go/middleware"
	"betelite-go/services"
	"betelite-go/utils"
)

func SetupBetRoutes(api fiber.Router) {
	bets := api.Group("/bets", middleware.AuthRequired())

	bets.Post("/place", func(c *fiber.Ctx) error {
		var req struct {
			MatchID string  `json:"matchId"`
			Pick    string  `json:"pick"` // home, away, draw
			Odds    float32 `json:"odds"`
			Amount  int64   `json:"amount"` // Note: kobo/pesewas
		}
		if err := c.BodyParser(&req); err != nil {
			return utils.SendError(c, 400, "Invalid payload")
		}

		if req.Amount <= 0 {
			return utils.SendError(c, 400, "Amount must be greater than 0")
		}

		// Verify match is live and taking bets
		match := services.Engine.GetMatch(req.MatchID)
		if match == nil {
			return utils.SendError(c, 400, "Match not available for betting")
		}
		if match.Minute >= 80 {
			return utils.SendError(c, 400, "Betting is closed for this match (80+ mins)")
		}

		uid := middleware.GetUID(c)
		ctx := context.Background()

		// DB Transaction
		tx, err := db.Pool.Begin(ctx)
		if err != nil {
			return utils.SendError(c, 500, "Database error")
		}
		defer tx.Rollback(ctx)

		// Check balance
		var balance int64
		err = tx.QueryRow(ctx, "SELECT balance FROM users WHERE id = $1 FOR UPDATE", uid).Scan(&balance)
		if err != nil || balance < req.Amount {
			return utils.SendError(c, 400, "Insufficient funds")
		}

		// We create the bet first so we have the betID for the transaction log
		betID := fmt.Sprintf("bet_%d", time.Now().UnixNano())
		potentialWin := int64(float64(req.Amount) * float64(req.Odds))

		_, err = tx.Exec(ctx, `
			INSERT INTO bets (id, user_id, match_id, pick, odds, amount, potential_win, status)
			VALUES ($1, $2, $3, $4, $5, $6, $7, 'live')`,
			betID, uid, req.MatchID, req.Pick, req.Odds, req.Amount, potentialWin)
		if err != nil {
			return utils.SendError(c, 500, "Failed to place bet")
		}

		err = services.AdjustBalance(ctx, tx, uid, -req.Amount, "bet_place", betID)
		if err != nil {
			return utils.SendError(c, 500, "Failed to update balance")
		}

		if err := tx.Commit(ctx); err != nil {
			return utils.SendError(c, 500, "Transaction commit failed")
		}

		return utils.SendSuccess(c, fiber.Map{"betId": betID})
	})
}
